onnx/defs/schema.h

```diff
@@ -33,6 +33,15 @@ typedef std::unordered_set<DataType> DataTypeSet;
 typedef std::unordered_map<std::string, std::pair<DataTypeSet, std::string>>
     TypeConstraintMap;
 
+typedef TensorShapeProto_Dimension InferenceDimension;
+
+struct InferenceContext {
+  virtual const AttributeProto* getAttribute(std::string name) = 0;
+  virtual const std::vector<TypeProto_Tensor>& getAllInputTypes() = 0;
+};
+
+typedef std::vector<TypeProto_Tensor> (*InferenceFunction)(InferenceContext&);
```

Comment:
if putting the output type also as an interface of InferenceContext, then the signature here could be changed to 

typedef bool (*InferenceFunction)(InferenceContext&); return value indicates whether shape inference done successfully or not.

Reply:
what types of failure would you like to signal? I think it should not be required that all output shapes are fully inferred.

Reply:
there may be cases that output shapes can't be inferred I think. In those cases, we may want to return "false" to callers to indicate that. Ideally, it would be better to have a "Status" class representing both flag (true|false) and some error message (std::string) if any. Make sense please?

Reply:
@bddppq are you ok with this? Since you initially asked for the return type to be the vector of output types itself.